atomicfu/src/nativeMain/kotlin/kotlinx/atomicfu/Synchronized.kt

```diff
@@ -0,0 +1,204 @@
+package kotlinx.atomicfu
+
+import platform.posix.*
+import interop.*
+import kotlinx.cinterop.*
+import kotlin.native.concurrent.AtomicNativePtr
+import kotlin.native.concurrent.AtomicReference
+import kotlin.native.concurrent.SharedImmutable
+import kotlin.native.concurrent.freeze
+import kotlin.native.internal.NativePtr
+import kotlinx.atomicfu.Status.*
+
+public actual open class SynchronizedObject {
+
+    private val lock = AtomicReference(LockState(UNLOCKED, 0, 0))
+
+    public fun lock() {
+        val currentThreadId = pthread_self()
+        while (true) {
+            val state = lock.value
+            when (state.status) {
+                UNLOCKED -> {
+                    val thinLock = LockState(THIN, 1, 0, currentThreadId)
+                    if (lock.compareAndSet(state, thinLock)) return
+                }
+                THIN -> {
+                    if (currentThreadId == state.ownerThreadId) {
+                        // reentrant lock
+                        val thinNested = LockState(THIN, state.nestedLocks + 1, state.waiters, currentThreadId)
+                        if (lock.compareAndSet(state, thinNested)) return
+                    } else {
+                        // another thread is trying to take this lock -> allocate native mutex
+                        val mutex = mutexPool.allocate(locked = true)
```

Comment:
It don't see why you need to pass `locked = true` to `mutexPool.allocate`. You can just call `mutexPool.allocate()` and call `mutex.lock()` here right after `allocate()`. Moreover, you don't have any other calls to `mutexPool.allocate()`, so this whole boolean is an extra code anyway.